src/validation.cpp

```diff
@@ -465,6 +464,111 @@ static bool CheckInputsFromMempoolAndCache(const CTransaction& tx, TxValidationS
 }
 
+/**
+ * A CoinsView that adds a memory cache to another CoinsView and serves as temporary scratch space.
+ * Used by MemPoolAccept class to validate transactions and packages before submitting to mempool.
+ * A backend can be set to provide read access to chainstate and/or mempool coins, but writing to
+ * the backend is disabled. Avoid using a CoinsViewTemporary in consensus-critical paths such
+ * as writing to the script cache. See CheckInputsFromMempoolAndCache as an example. When not being
+ * used to validate a package (m_temp_added and m_temp_spent are empty), a CoinsViewTemporary
+ * behaves exactly like a CCoinsViewCache.
+ */
+class CoinsViewTemporary : public CCoinsViewCache
+{
+    /**
+    * Coins made available by transactions being validated. Tracking these allows for package
+    * validation, since we can access transaction outputs without submitting them to mempool.
+    */
+    std::map<COutPoint, Coin> m_temp_added;
+
+    /**
+    * Coins spent by transactions being validated. When validating a package, we need to track
+    * these in order to distinguish between missing/spent coins and conflicts within a package.
+    */
+    std::set<COutPoint> m_temp_spent;
+
+    /** An empty coin used as a placeholder for a spent coin.*/
+    inline static const Coin s_coin_spent;
+
+public:
+
+    CoinsViewTemporary(CCoinsView* baseIn) : CCoinsViewCache(baseIn) {}
+
+    // Delete the copy constructor to prevent accidentally using it when one intends to create a
+    // CoinsViewTemporary on top of a base cache.
+    CoinsViewTemporary(const CoinsViewTemporary &) = delete;
+
+    bool GetCoin(const COutPoint& outpoint, Coin& coin) const override {
+        coin = AccessCoin(outpoint);
+        return !coin.IsSpent();
+    }
+
+    const Coin& AccessCoin(const COutPoint& outpoint) const override {
+        // Check to see if another tx in the package has already spent this coin (conflict-in-package).
+        // Coins spent by others in the package are only tracked in m_temp_spent.
+        if (m_temp_spent.count(outpoint)) {
+            return s_coin_spent;
+        }
+
+        // Check to see if the inputs are made available by another tx in the package.
+        // These Coins would not be available in the underlying CoinsView.
+        if (auto it = m_temp_added.find(outpoint); it != m_temp_added.end()) {
+            assert(!it->second.IsSpent());
+            return it->second;
+        }
+        return CCoinsViewCache::AccessCoin(outpoint);
+    }
+
+    bool HaveCoin(const COutPoint& outpoint) const override {
+        Coin coin;
+        return GetCoin(outpoint, coin);
+    }
+
+    /**
+    * Update with coins spent and created by a transaction.
+    * Only used for package validation.
+    */
+    void PackageAddTransaction(const CTransactionRef& tx)
+    {
+        // Track Coins spent by this transaction. They must exist and not already be spent.
```

Comment:
this should be documented in the function doc that this must be true.

Reply:
Done